tcms/core/ajax.py

```diff
@@ -194,105 +191,76 @@ class Objects(object):
-                    self.object = o
-                    self.object_pks = request.GET.getlist(o)
+            for obj in self.__all__:
+                if request.GET.get(obj):
+                    self.object = obj
+                    self.object_pk = request.get(obj)
                     break
 
         def get(self):
             func = getattr(self, self.object)
             return func()
 
         def plan(self):
-            return self.template_name, TestPlan.objects.filter(
-                pk__in=self.object_pks)
+            return self.template_name, TestPlan.objects.get(
+                pk=self.object_pk)
 
         def case(self):
-            return self.template_name, get_selected_testcases(self.request)
+            return self.template_name, TestCase.objects.get(
+                pk=self.object_pk)
 
         def run(self):
-            self.template_name = 'run/get_tag.html'
-            return self.template_name, TestRun.objects.filter(
-                pk__in=self.object_pks)
+            return 'run/get_tag.html', TestRun.objects.get(
+                pk=self.object_pk)
 
     class TagActions(object):
         __all__ = ['add', 'remove']
 
         def __init__(self, obj, tag):
             self.obj = obj
-            self.tag = string_to_list(tag)
-            self.request = request
+            self.tag = tag
 
         def add(self):
-            for tag_str in self.tag:
-                try:
-                    tag, c = Tag.objects.get_or_create(name=tag_str)
-                    for o in self.obj:
-                        o.add_tag(tag)
-                except IntegrityError:
-                    return "Error when adding %s " \
-                           "Tag with this key already exists in the database." % self.tag
-                except Exception:
-                    return "Error when adding %s" % self.tag
+            tag, _ = Tag.objects.get_or_create(name=self.tag)
+            self.obj.add_tag(tag)
 
             return True, self.obj
 
         def remove(self):
-            self.obj = self.obj.filter(tag__name__in=self.tag).distinct()
-
-            if not self.obj:
-                return "Tags does not exist in current selected plan."
-
-            for tag_str in self.tag:
-                try:
-                    tag = Tag.objects.filter(name=tag_str)[0]
-                except IndexError:
-                    return "Tag %s does not exist in current selected plan." % tag_str
-                for object in self.obj:
-                    try:
-                        object.remove_tag(tag)
-                    except Exception:
-                        return "Remove tag %s error." % tag
+            tag = Tag.objects.get(name=self.tag)
+            self.obj.remove_tag(tag)
+
             return True, self.obj
 
     objects = Objects(request, template_name)
     template_name, obj = objects.get()
 
     q_tag = request.GET.get('tags')
     q_action = request.GET.get('a')
-    if q_action:
```

Comment:
You will have to return this if statement and also add a test for it or update existing ones. 

When rendering tags for TestCase or TestPlan the UI calls this same view function but without the 'a=XXX' query string parameter in the URL. Before the change this resulted in not calling methods from TagActions but still rendering the template. Now it will break because q_action will be None and you try to access a methods based on the value of this variable. This has to do with the rendering I mention in my other comment so maybe the longer term solution will be to split the add/remove actions from the get+rendering action.